### PR TITLE
Support authentication for the OssIndex advisor

### PR DESCRIPTION
**Pull request by sschuberth**
Please have a look at the individual commit messages for the details.

### CODE REVIEW

**Review by tsteenbe — CHANGES_REQUESTED** [review]
@sschuberth Could you include in this PR an update of [Advisor in reference.yml]( https://github.com/oss-review-toolkit/ort/blob/main/model/src/main/resources/reference.yml#L104) so ORT users can easily find out authentication for the OssIndex is supported?

File: advisor/src/main/kotlin/advisors/OssIndex.kt [hunk 1]
```diff
@@ -64,11 +64,18 @@ class OssIndex(name: String, config: OssIndexConfiguration) : AdviceProvider(nam
...
-            url = config.serverUrl,
-            client = OkHttpClientHelper.buildClient()
+            config.serverUrl,
+            config.username,
+            config.password,
+            OkHttpClientHelper.buildClient()
         )
     }
 
+    private val getComponentReport by lazy {
```
Comment (fviernau) [inline advisor/src/main/kotlin/advisors/OssIndex.kt:74]: What do you think about explicitly specifying the type of `getComponentReport()` for readability? 

Alternatively, this performance optimization can be dropped in favor of a (IMO more readable) function. To me it sacrifices readability without an obvious need.

What do you think?

**Review by fviernau — CHANGES_REQUESTED** [review]

File: advisor/src/main/kotlin/advisors/OssIndex.kt (hunk 1 above)
Comment (sschuberth) [reply advisor/src/main/kotlin/advisors/OssIndex.kt:74]: > What do you think?

I think it's not less readable than a wrapping function, and (letting the performance aspect aside) on the contrary more elegant.

Adding an explicit type would actually *reduce* readability, as this is a function reference to a `suspend` function, so the type would be something like `KSuspendFunction1<ComponentReportRequest, List<ComponentReport>>`.

Finally, using a function would require to pass `config` as an argument, as functions do not have access to constructor arguments like property initialization has.

So all in all, I'd prefer to keep it as-is.

*Edit:* Here the direct comparison

```
private val getComponentReport by lazy {
    val hasCredentials = config.username != null && config.password != null
    if (hasCredentials) service::getAuthorizedComponentReport else service::getComponentReport
}
```

vs

```
private suspend fun getComponentReport(config: OssIndexConfiguration, components: ComponentReportRequest): List<ComponentReport> {
    val hasCredentials = config.username != null && config.password != null
    return if (hasCredentials) service.getAuthorizedComponentReport(components) else service.getComponentReport(components)
}
```

to me the former is less repetitive, more elegant, and thus more readable and better.

Comment (fviernau) [reply advisor/src/main/kotlin/advisors/OssIndex.kt:74]: I had something like below in mind, however - let's just leave it as is.

```
private suspend fun ComponentReportRequest.execute(): List<ComponentReport> =
   when {
        config.hasCredentials() -> service.getAuthorizedComponentReport(this)
        else -> service.getComponentReport(this)
   }

private fun OssIndexConfiguration.hasCredentials() = config.username != null && config.password != null
```

**Review by fviernau — APPROVED** [review]

File: advisor/src/main/kotlin/advisors/OssIndex.kt (hunk 1 above)
Comment (sschuberth) [reply advisor/src/main/kotlin/advisors/OssIndex.kt:74]: I like the extension function, though. Let me add that in a follow-up PR.

**Pull request merged**